mawie81/electron-window-state
# Fix: Validate state when saved display unavailable

## What is this PR

This PR fixes the window positioning at startup when the saved display is unavailable.

See the corresponding issue: https://github.com/mawie81/electron-window-state/issues/44.
Related: https://github.com/mawie81/electron-window-state/issues/1, https://github.com/mawie81/electron-window-state/issues/4, https://github.com/mawie81/electron-window-state/issues/27, https://github.com/mawie81/electron-window-state/issues/31, https://github.com/mawie81/electron-window-state/issues/38, https://github.com/mawie81/electron-window-state/issues/43.

Commit aac92c0 shows that the current implementation is broken since the last test is failing (`Ensure window is visible at startup if saved display is unavailable and was on the left`).

## How does it work
The solution currently adopted consists of reseting the state if the saved display is unavailable.

The advantage is that we ensure the window will always be visible if the saved display is not found. The downside is that the window will not be positioned to the saved coordinates if another display contains the corresponding bounds.

Better implementations that check for each display available and ensure the window is fully visible on one can be found in `electron-boilerplate` and `VS Code`.
See:
- [electron-boilerplate:src/helpers/window.js](https://github.com/szwacz/electron-boilerplate/blob/bd4812720b0bd5f21beb413e567723bae653d5f5/src/helpers/window.js#L58-L68)
- [vscode:src/vs/code/electron-main/window.ts](https://github.com/Microsoft/vscode/blob/95ca89baf72f1e6b3db528533d8bd731e0430103/src/vs/code/electron-main/window.ts#L649-L705)

What about using their methods?

## Test instructions
- Fetch this branch and run `npm link`
- Clone https://github.com/MathieuDebit/electron-quick-start/pull/3
- Run `npm link electron-window-state`
- Follow the reproductions steps on the [pull request](https://github.com/MathieuDebit/electron-quick-start/pull/3) and verify that the bug is fixed

## Todos
